File: app/src/main/java/parkhomov/andrew/lensthicknesscalculator/activities/MainActivity.java
package parkhomov.andrew.lensthicknesscalculator.activities;

import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.content.ContextCompat;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Gravity;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.Toast;

import java.util.Locale;

import parkhomov.andrew.lensthicknesscalculator.R;
import parkhomov.andrew.lensthicknesscalculator.customViews.ScrimInsetsFrameLayout;
import parkhomov.andrew.lensthicknesscalculator.fragments.DiamCalculatorFragment;
import parkhomov.andrew.lensthicknesscalculator.fragments.ThknsCalculatorFragment;
import parkhomov.andrew.lensthicknesscalculator.utils.UtilsDevice;
import parkhomov.andrew.lensthicknesscalculator.utils.UtilsMiscellaneous;

/**
 * Main activity class. Customize drawers, toolbar, fragment behaviour ect.
 */
public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    public static boolean isTextLinkInDatabaseClicked;
    private DrawerLayout drawerLayout;
    private LinearLayout navDrawerEntriesRootView;
    private FrameLayout frameLayoutThknsCalc, frameLayoutDiamCalc,
            frameLayoutSettings, frameLayoutAbout,frameLayoutGlossary;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setUpLanguage();
        setContentView(R.layout.activity_main);
        initialize();
    }

    private void setUpLanguage() {
        LanguageManagerActivity languageManager = new LanguageManagerActivity(this);
        languageManager.setCurrentLanguage();
    }

    private void initialize() {
        final Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        if (getSupportActionBar() != null){
            getSupportActionBar().setTitle(getString(R.string.string_thkns_calc));
        }

        setUpIcons(R.id.navigation_drawer_items_list_icon_thkns_calc);
        setUpIcons(R.id.navigation_drawer_items_list_icon_diam_calc);
        setUpIcons(R.id.navigation_drawer_items_list_icon_glossary);

        navDrawerEntriesRootView = (LinearLayout) findViewById
                (R.id.navigation_drawer_linearLayout_entries_root_view);

        frameLayoutThknsCalc = (FrameLayout) findViewById
                (R.id.navigation_drawer_items_list_linearLayout_thkns_calc);

        frameLayoutDiamCalc = (FrameLayout) findViewById
                (R.id.navigation_drawer_items_list_linearLayout_diamCalc);

        frameLayoutAbout = (FrameLayout) findViewById
                (R.id.navigation_drawer_items_list_linearLayout_about);

        frameLayoutSettings = (FrameLayout) findViewById
                (R.id.navigation_drawer_items_list_linearLayout_settings);

        frameLayoutGlossary = (FrameLayout) findViewById
                (R.id.navigation_drawer_items_linearLayout_glossary);

        // Navigation Drawer
        drawerLayout = (DrawerLayout) findViewById(R.id.main_activity_DrawerLayout);

        final ScrimInsetsFrameLayout mScrimInsetsFrameLayout = (ScrimInsetsFrameLayout)
                findViewById(R.id.main_activity_navigation_drawer_rootLayout);

        final ActionBarDrawerToggle mActionBarDrawerToggle = new ActionBarDrawerToggle(
                this,
                drawerLayout,
                toolbar,
                R.string.navigation_drawer_opened,
                R.string.navigation_drawer_closed) {
            @Override
            public void onDrawerSlide(View drawerView, float slideOffset) {
                // Disables the burger/arrow animation by default
                super.onDrawerSlide(drawerView, 0);
            }
        };
        drawerLayout.openDrawer(Gravity.LEFT);
        drawerLayout.setDrawerListener(mActionBarDrawerToggle);

        // Navigation Drawer layout width
        final int possibleMinDrawerWidth =
                UtilsDevice.getScreenWidth(this) -
                        UtilsMiscellaneous
                                .getThemeAttributeDimensionSize(this, android.R.attr.actionBarSize);

        final int maxDrawerWidth =
                getResources().getDimensionPixelSize(R.dimen.navigation_drawer_max_width);

        mScrimInsetsFrameLayout.getLayoutParams().width =
                Math.min(possibleMinDrawerWidth, maxDrawerWidth);

        // Nav Drawer item click listener

        frameLayoutThknsCalc.setOnClickListener(this);
        frameLayoutDiamCalc.setOnClickListener(this);
        frameLayoutGlossary.setOnClickListener(this);
        frameLayoutAbout.setOnClickListener(this);
        frameLayoutSettings.setOnClickListener(this);

        Fragment fragment;
        String title;
        if(isTextLinkInDatabaseClicked){
            fragment = new DiamCalculatorFragment();
            title = getString(R.string.string_diam_calc);
            if(drawerLayout.isDrawerOpen(GravityCompat.START))drawerLayout.closeDrawer(Gravity.LEFT);
            frameLayoutDiamCalc.setSelected(true);
            isTextLinkInDatabaseClicked = false;
        }else{
            fragment = new ThknsCalculatorFragment();
            title = getString(R.string.string_thkns_calc);
            frameLayoutThknsCalc.setSelected(true);
        }
        FragmentManager fragmentManager = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        fragmentTransaction.replace(R.id.main_activity_content_frame, fragment);
        fragmentTransaction.addToBackStack(null);
        fragmentTransaction.commit();
        mActionBarDrawerToggle.syncState();
        if(getSupportActionBar() != null) {
            getSupportActionBar().setTitle(title);
        }
    }


    @Override
    public void onClick(View view) {
        if (!view.isSelected()) {
            onRowPressed((FrameLayout) view);
            Fragment fragment = null;
            String title = null;
            if (view == frameLayoutThknsCalc) {
                fragment = new ThknsCalculatorFragment();
                title = getString(R.string.string_thkns_calc);
            } else if (view == frameLayoutDiamCalc) {
                fragment = new DiamCalculatorFragment();
                title = getString(R.string.string_diam_calc);
            }
            if (fragment != null){
                FragmentManager fragmentManager = getSupportFragmentManager();
                FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
                fragmentTransaction.replace(R.id.main_activity_content_frame, fragment);
                fragmentTransaction.addToBackStack(null);
                fragmentTransaction.commit();
                view.setSelected(true);
                if (getSupportActionBar() != null){
                    getSupportActionBar().setTitle(title);
                }
            }
            if (view == frameLayoutGlossary) {
                startActivity(new Intent(view.getContext(), GlossaryListActivity.class));
            } else if (view == frameLayoutAbout) {
                startActivity(new Intent(view.getContext(), AboutDialogActivity.class));
            } else if (view == frameLayoutSettings) {
                startActivity(new Intent(view.getContext(), SettingsActivity.class));
            }
        } else {
            drawerLayout.closeDrawer(GravityCompat.START);
        }
    }

    private void onRowPressed(@NonNull final FrameLayout pressedRow) {
        if (pressedRow.getTag() != getResources().getString(R.string.tag_nav_drawer_special_entry))
        {
            for (int i = 0; i < navDrawerEntriesRootView.getChildCount(); i++)
            {
                final View currentView = navDrawerEntriesRootView.getChildAt(i);

                final boolean currentViewIsMainEntry = currentView.getTag() ==
                        getResources().getString(R.string.tag_nav_drawer_main_entry);

                if (currentViewIsMainEntry) {
                    currentView.setSelected(currentView == pressedRow);
                }
            }
        }

        drawerLayout.closeDrawer(GravityCompat.START);
    }

    private void setUpIcons(int imageId) {
        // Icons tint list
        final ImageView imageView =
                (ImageView) findViewById(imageId);
        if(imageView != null){
            final Drawable drawable = DrawableCompat.wrap(imageView.getDrawable());
            DrawableCompat.setTintList(
                    drawable.mutate(),
                    ContextCompat.getColorStateList(this, R.color.nav_drawer_icon)
            );
            imageView.setImageDrawable(drawable);
        }
    }

    @Override
    public void onBackPressed() {
        if(drawerLayout.isDrawerOpen(GravityCompat.START)){
            drawerLayout.closeDrawer(Gravity.LEFT);
        }else{
            new AlertDialog.Builder(this)
                    .setTitle(R.string.exit_question_title)
                    .setMessage(R.string.exit_question)
                    .setNegativeButton(android.R.string.no, null)
                    .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {

                        public void onClick(DialogInterface arg0, int arg1) {
                            android.os.Process.killProcess(android.os.Process.myPid());
                        }
                    }).create().show();
        }
    }
}

File: app/src/main/java/parkhomov/andrew/lensthicknesscalculator/activities/main/MainActivity.kt
package parkhomov.andrew.lensthicknesscalculator.activities.main

import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.support.design.widget.TabLayout
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import android.support.v4.view.ViewPager
import android.support.v7.app.AlertDialog
import android.view.View
import android.widget.TextView
import android.widget.Toast

import java.util.ArrayList

import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import parkhomov.andrew.lensthicknesscalculator.R
import parkhomov.andrew.lensthicknesscalculator.R2
import parkhomov.andrew.lensthicknesscalculator.activities.fragment.glossary.GlossaryDetails
import parkhomov.andrew.lensthicknesscalculator.activities.fragment.glossary.GlossaryList
import parkhomov.andrew.lensthicknesscalculator.activities.fragment.settings.Language
import parkhomov.andrew.lensthicknesscalculator.activities.fragment.settings.Settings
import parkhomov.andrew.lensthicknesscalculator.activities.interfaces.LanguageChangedI
import parkhomov.andrew.lensthicknesscalculator.activities.tabs.AbstractTabFragment
import parkhomov.andrew.lensthicknesscalculator.activities.tabs.Diameter
import parkhomov.andrew.lensthicknesscalculator.activities.tabs.TabsPageFragmentAdapter
import parkhomov.andrew.lensthicknesscalculator.activities.tabs.Thickness
import parkhomov.andrew.lensthicknesscalculator.activities.utils.CONSTANT
import parkhomov.andrew.lensthicknesscalculator.activities.utils.Utils
import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper

/**
 * Main activity class. Customize drawers, toolbar, fragment behaviour ect.
 */
class MainActivity : FragmentActivity(), LanguageChangedI {

    interface HideKeyboardI {
        fun hideKeyboard()
    }

    @BindView(R2.id.viewPager)
    lateinit var viewPager: ViewPager
    @BindView(R2.id.MyTabLayout)
    lateinit var tabLayout: TabLayout
    @BindView(R2.id.header)
    lateinit var header: TextView

    private val headers = ArrayList<String>(12)
    private val description = ArrayList<String>(12)
    private val images = ArrayList<Int>(12)

    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_activity)
        ButterKnife.bind(this)
        createListWithData()
        createTabs()
    }

    private fun createTabs() {
        val tabsPageFragmentAdapter = TabsPageFragmentAdapter(
                supportFragmentManager,
                headers,
                description,
                images
        )
        viewPager!!.offscreenPageLimit = 2
        viewPager!!.adapter = tabsPageFragmentAdapter
        tabLayout!!.setupWithViewPager(viewPager)

        viewPager!!.setPageTransformer(false, Utils.NoPageTransformer())
    }

    override fun onBackPressed() {
        val glossaryList = supportFragmentManager.findFragmentByTag(CONSTANT.GLOSSARY_LIST) as GlossaryList
        val glossaryDetails = supportFragmentManager.findFragmentByTag(CONSTANT.GLOSSARY_DETAILS) as GlossaryDetails
        val language = supportFragmentManager.findFragmentByTag(CONSTANT.LANGUAGE) as Language
        val settings = supportFragmentManager.findFragmentByTag(CONSTANT.SETTINGS) as Settings

        backPressedBehaviour(
                glossaryList,
                glossaryDetails,
                language,
                settings
        )

    }

    private fun backPressedBehaviour(vararg fragment: Fragment) {
        val isFragmentAdded = fragment.any { it.isAdded }
        // check if one of one fragment is added

        if (isFragmentAdded) {
            supportFragmentManager.popBackStack()
        } else {
            AlertDialog.Builder(this)
                    .setTitle(R.string.exit_question_title)
                    .setMessage(R.string.exit_question)
                    .setNegativeButton(android.R.string.no, null)
                    .setPositiveButton(android.R.string.yes) { arg0, arg1 -> android.os.Process.killProcess(android.os.Process.myPid()) }.create().show()
        }
    }

    @OnClick(R2.id.openGlossary)
    fun onGlossaryClicked() {
        try {
            supportFragmentManager
                    .beginTransaction()
                    .addToBackStack(null)
                    .add(R.id.mainContainerConstr, GlossaryList.getInstance(headers, description, images), CONSTANT.GLOSSARY_LIST)
                    .commit()
        } catch (e: IllegalStateException) {
        }

        closeSoftKeyboard()
    }

    @OnClick(R.id.openSettings)
    fun onStatisticClicked() {
        try {
            supportFragmentManager
                    .beginTransaction()
                    .addToBackStack(null)
                    .add(R.id.mainContainerConstr, Settings.getInstance(this), CONSTANT.SETTINGS)
                    .commit()
        } catch (e: IllegalStateException) {
        }

        closeSoftKeyboard()
    }

    /* try to hide all soft keyboards in viewpager. One of tab is not in focus and we get 100%
    ClassCastException, and we handle it and ignore */
    private fun closeSoftKeyboard() {
        val index = viewPager.currentItem

        val adapter = viewPager.adapter as TabsPageFragmentAdapter

        try {
            val hide = adapter.getFragment(index) as Thickness
            hide.hideKeyboard()
        } catch (e: ClassCastException) {
        } catch (e: NullPointerException) {
        }

        try {
            val hide = adapter.getFragment(index) as Diameter
            hide.hideKeyboard()
        } catch (e: ClassCastException) {
        } catch (e: NullPointerException) {
        }

    }

    // this code reload activity when user change language
    override fun languageChanged() {
        finish()
        overridePendingTransition(R.anim.fade_in, R.anim.fade_out)
        startActivity(intent)
    }

    private fun createListWithData() {
        //add headers
        headers.add(0, getString(R.string.index_of_refraction))
        headers.add(1, getString(R.string.sphere_power))
        headers.add(2, getString(R.string.cylinder_power))
        headers.add(3, getString(R.string.axis))
        headers.add(4, getString(R.string.real_base_curve))
        headers.add(5, getString(R.string.center_thickness))
        headers.add(6, getString(R.string.edge_thickness))
        headers.add(7, getString(R.string.diameter))
        headers.add(8, getString(R.string.effective_diameter))
        headers.add(9, getString(R.string.distance_between_lenses))
        headers.add(10, getString(R.string.pupil_distance))
        headers.add(11, getString(R.string.transposition))
        // add description
        description.add(0, getString(R.string.description_index_of_refraction))
        description.add(1, getString(R.string.description_sphere_power))
        description.add(2, getString(R.string.description_cylinder_power))
        description.add(3, getString(R.string.description_axis))
        description.add(4, getString(R.string.description_real_base_curve))
        description.add(5, getString(R.string.description_center_thickness))
        description.add(6, getString(R.string.description_edge_thickness))
        description.add(7, getString(R.string.description_diameter))
        description.add(8, getString(R.string.description_effective_diameter))
        description.add(9, getString(R.string.description_distance_between_lenses))
        description.add(10, getString(R.string.description_pupil_distance))
        description.add(11, getString(R.string.description_transposition))
        // images for each item
        images.add(0, R.drawable.index_of_refraction_img)
        images.add(1, R.drawable.sphere_img)
        images.add(2, R.drawable.cylinder_img)
        images.add(3, R.drawable.axis_img)
        images.add(4, R.drawable.front_curve_img)
        images.add(5, R.drawable.thickness_gauge_img)
        images.add(6, R.drawable.edge_thickness_img)
        images.add(7, R.drawable.diam_img)
        images.add(8, R.drawable.ed_img)
        images.add(9, R.drawable.dbl_img)
        images.add(10, R.drawable.pd_img)
        images.add(11, R.drawable.transposition_img)
    }
}
